Data/HashMap/Internal.hs

```diff
@@ -1771,37 +1768,135 @@ differenceWith f a b = foldlWithKey' go empty a
 -- | /O(n*log m)/ Intersection of two maps. Return elements of the first
 -- map for keys existing in the second.
 intersection :: (Eq k, Hashable k) => HashMap k v -> HashMap k w -> HashMap k v
-intersection a b = foldlWithKey' go empty a
-  where
-    go m k v = case lookup k b of
-                 Just _ -> unsafeInsert k v m
-                 _      -> m
+intersection = intersectionWith const
 {-# INLINABLE intersection #-}
```

Comment:
I think we probably want to explicitly `inline intersectionWithKey` here. We really don't want stupid thunks in the result. Another possibility might be to base them all (lazy and strict) on a version that takes `k -> v -> v -> (# v #)`, though that could be a bit slower.

Reply:
I would favor using unboxed tuples. What are some reasons why it would be slower?